Refactor the REST API

Xebax opened the pull request:
Refactor the REST API:

1) Modify the definition of the API so that it is closest to the commonly accepted practices. For example:
- get  the list of third parties: GET /thirdparty/list is replaced by GET /thirdparties
- create a third party: POST /thirdparty is replaced by POST /thirdparties

2) Use the Restler API Explorer for documenting the API. It allows browsing the API and also playing with it, which is a good way to learn how to use it.

Xebax committed:
REST API: enable the Restler API Explorer.

Xebax committed:
REST API: begin the refactoring.

Simplify the code that looks for the API classes and replace the
/contact/ and /user/ APIs by /contacts/ and /users/.

Xebax committed:
REST API: replace the /category/ API by /categories/.

Remove several /category/list/xxx endpoints because they are redundant
with the /categories/ endpoint whith the 'type' filter.

Xebax committed:
REST API: replace the /invoice/ API by /invoices/.

Rename the 'mode' parameter to 'status' in GET /invoices.
Fix the error handling in DELETE /invoices/{id}.

Xebax committed:
REST API: replace the /product/ API by /products/.

Added some FIXME:
 - the put() method does its job but the API returns an error 500
 - delete() does not work.

Xebax committed:
REST API: replace the /order/ API by /orders/.

Remove the /thirdparty/{socid}/order/list endpoint because it is
redundant with the /orders/ endpoint whith the 'societe' filter.

Xebax committed:
REST API: replace the /thirdparty/ API by /thirdparties/.

Remove several /thirdparty/list/xxx endpoints because they are redundant
with the /thirdparties/ endpoint whith the 'mode' filter.
Remove the customer/byemail/{email} endpoint because it is redundant
with the /thirdparties/ endpoint whith the 'email' filter.
Remove all the /customer/ endpoints because they do the same thing as
the corresponding /thirdparty/ endpoints.

Xebax committed:
REST API: update the API to have /login and /status.

Replace the GenericApi class by two classes, Login and Status, else the
endpoints would be /generic/login and /generic/status.

Xebax committed:
REST API: sort the classes before adding them to Restler.

The Restler API Explorer shows the classes in the order they are added
and it's a mess if they are not sorted.

Xebax committed:
REST API: fix the error 500 when updating a product.

Xebax committed:
REST API: fix deleting a product/service.

Add a User parameter to the Product::delete() method to explicitly set
the user that performs the action instead of using a global variable.
In the calls to Product::delete() with a fetched object, remove the
deprecated parameter $id.

Xebax committed:
REST API: move the definition of /contacts/{id}/createUser
from api_users.class.php to api_contacts.class.php.

Xebax committed:
REST API: add the 'category' filter in /products/

and fix the 'to_sell' and 'to_buy' filters.

Xebax committed:
REST API: remove getByCategory() from Products

because the same feature is available with GET /products/.

Xebax committed:
REST API: fix the definition of /orders/{id}/validate.

Xebax committed:
REST API: improve the API to get/add categories for an item.

Move the GET /xxx/{id}/categories requests to the corresponding classes
(Contacts, Products, Thirdparties).
Move the addCustomerCategory() method to the Thirdparties class. Use the
POST verb for the request instead of GET because the request modifies
the state of the server.

Xebax committed:
REST API: update the URLs in RestAPIUserTest.php.

Xebax committed:
REST API: fix the index() method in Categories.

In the Categories class, set all the parameters of the index() method as
optional to make this method compatible with the one of the parent
class, DolibarrApi.

Xebax committed:
REST API: replace the home made API explorer by the Restler API Explorer.

jfefe commented:
As said @Eldy the explorer sould be into an external module.

jfefe commented:
To have new url for API, you just need to add doc string for it. ie : 

```
@url    GET /category/list
@url    GET /categories
```

Those 2 url will be available into API without changing anything. Please keep compatibility with changed endpoint by using doc string. 

/thirparties for POST method is a discutable choice :) With this I would tend to think that we can create several thirparties.

fappels commented on htdocs/product/class/product.class.php:
```diff
@@ -878,18 +878,19 @@ function update($id, $user, $notrigger=false, $action='update')
 	 * 	@return		int					< 0 if KO, 0 = Not possible, > 0 if OK
 	 */
-	function delete($id=0, $notrigger=0)
+	function delete($user, $id=0, $notrigger=0)
```
This will break dolibarr. If you want to remove deprecated id parameter, you should replace it by the user object and ignore first parameter if int. Also better put this is separate PR.

Xebax replied:
@fappels Thank you for your feedback. Yes, you are right. I was not sure of myself when I did these changes. This is the only commit that modifies code outside of the REST API. I will revert it and fix the bug another way.

Xebax committed:
Revert "REST API: fix deleting a product/service."

This reverts commit d7c8a466b12c99306a2080cd22d887a36c9f751d.

Xebax committed:
REST API: fix deleting a product/service.

The global variable $user must be set before calling the
Product::delete() method else the check of permissions fails.

Xebax commented:
@jfefe @eldy I don't understand the comment about the explorer that should be in an external module.
Dolibarr uses Restler, the code of the framework is included in Dolibarr, and the Restler API Explorer is contained in Restler. Using the Restler API Explorer only requires adding this line in the code (in htdocs/api/index.php):
`$api->r->addAPIClass('Luracast\\Restler\\Explorer');`
and then the explorer is available at https://_yourdolibarr.tld/mydolibarr_/api/index.php/explorer

Xebax committed:
REST API: Bring back the existing API (for backward compatibility).

Bring back the existing API and mark all the operations as deprecated.
Both the existing and the new API are documented and browsable with the
Restler API Explorer. The existing API appears under "root".

Xebax committed:
REST API: Add comments in the @deprecated tags.

Xebax commented:
Here is a screenshot of the modified API browsed with the Restler API Explorer:
![dolibarr - modified rest api](https://cloud.githubusercontent.com/assets/8964954/16260447/d1bdf19c-3867-11e6-95f2-4071750d6ba7.png)

Please note that the "root" entry at the top is the one that contains the existing API, for backward compatibility.

rdoursenaud commented:
I think that the rationale for having the explorer in a separate module is to limit Dolibarr distribution size.
This would allow packaging Dolibarr without the explorer and offer the explorer as a downloadable module on Dolistore for those who need it.
Looks reasonable enough but can be a maintenance nightmare until we have an automated way of pushing modules to the Dolistore at release time.
Anyway, it could be a sound model to distribute modules that are not strictly needed in a basic Dolibarr install but still maintain them as part of the main repository.

My 2 cents…

Great work nonetheless!

Xebax commented:
@rdoursenaud I understand that distributing modules separately could be desirable. My point here is that I don't see how the Restler API Explorer - alone - could be provided in a separate module. The explorer is bundled in Restler and Restler is included in Dolibarr. It is mandatory for the REST API module.
If something must be distributed in a separate module, it would be the whole REST API, including Restler and the explorer. I do not see other option but I may have missed something.

eldy commented:
A standard distribution of dolibarr is a package for USER.
RESTLER "explorer" is a tool for developper or integrator. Currently, such tools are stored into "dev" directory and they are filtered (extracted) when building final package for user.
I will make another test with you suggestion of adding a line in index.php (atfer release of 4.0). If it does not need to add libraries that is required ONLY for the dev feature, and if it does not bypass dolibarr security and work out of the box, i will reconsider my choice

Xebax commented:
I don't understand the point about user VS developer. We are talking about web services here. The users of web services are developers, sure. But viewed from Dolibarr, they are still users: developing a client application that uses the web services of Dolibarr is not the same as developing inside of Dolibarr. I understand the usage of the "dev" for the internals of Dolibarr but here it is different.
When the SOAP web services are enabled, the WSDL files and the documentation are available in the configuration page.
It's the same for the REST web services: when they are available, the documentation (explorer) is available in the configuration page. My proposal is only to replace the home made explorer by the Restler API Explorer. Moreover, as explained above, it is bundled in the Restler framework, which is already included in Dolibarr. This PR does not contain any new library.

Please use this whole PR when you will make a test with the Restler API Explorer, else you will have an incomplete result.
I understand that the version 4.0 must be released before that, no problem with that.

eldy commented:
I finnally solve conflict manually

Xebax commented:
@eldy OK. Sorry, I got your message too late and you already had solved the conflict.
I see that the htdocs/api/admin/explorer.php file still exists after the merge but it is not needed anymore.

eldy commented:
Yes. I kept it because  ode it contains help me to make test and understanding the way explorer work with a very low need of bytes. I will decide later.
About root, do you means api you find here are different than api found into other topics ? If yes, is it possible to rename it i to oldapi ?
I also have an other question: do you know how to implement a field "modified since" on api that returns full list. Goal is to add a filter on field tms >= of the request filtering date to limit answers.
